beet/contrib/optifine.py

```diff
@@ -0,0 +1,67 @@
+"""Plugin for supporting optifine resources."""
+
+
+__all__ = [
+    "optifine",
+    "JsonEntityModel",
+    "JsonPartModel",
+    "OptifineProperties",
+    "ShaderProperties",
+    "OptifineTexture",
+]
+
+
+from typing import Union
+
+from beet import Context, JsonFile, ResourcePack, TextFile, Texture
+
+
+def beet_default(ctx: Context):
+    ctx.require(optifine)
+
+
+def optifine(pack: Union[Context, ResourcePack]):
+    """Enable optifine resources."""
+    if isinstance(pack, Context):
+        pack = pack.assets
+    pack.extend_namespace += [
+        JsonEntityModel,
+        JsonPartModel,
+        OptifineProperties,
+        ShaderProperties,
+        OptifineTexture,
+    ]
+
+
+class JsonEntityModel(JsonFile):
+    """Class representing a json entity model."""
+
+    scope = ("optifine", "cem")
+    extension = ".jem"
+
+
+class JsonPartModel(JsonFile):
+    """Class representing a json part model."""
+
+    scope = ("optifine", "cem")
+    extension = ".jpm"
+
+
+class OptifineProperties(TextFile):
+    """Class representing optifine properties."""
+
+    scope = ("optifine",)
+    extension = ".properties"
+
+
+class OptifineTexture(Texture):
```

Comment:
File types registered as namespaced resources should be treated as "final" classes (can't be inherited from). `Texture` is already registered for resource packs. Use `PngFile` instead (with an explicit `extension = ".png"` attribute).

In general, all classes ending with `-File` are meant to be abstract representations for a general file format. Subclasses that don't end with `-File` are concrete file types representing a specific usage of the parent file format in a data pack or resource pack.